cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,122 +1266,742 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+    }
+
+    auto& h_dst_buf_info  = partition_buf_size_and_dst_buf_info->h_dst_buf_info;
+    auto cur_dst_buf_info = h_dst_buf_info;
+    detail::metadata_builder mb{input.num_columns()};
+
+    populate_metadata(input.begin(), input.end(), cur_dst_buf_info, mb);
+
+    return std::make_unique<std::vector<uint8_t>>(std::move(mb.build()));
   }
-  return result;
+
+  std::vector<packed_table> make_packed_tables()
+  {
+    if (input.num_columns() == 0) { return std::vector<packed_table>(); }
+    if (is_empty) { return make_empty_packed_table(); }
+    std::vector<packed_table> result;
+    result.reserve(num_partitions);
+    std::vector<column_view> cols;
+    cols.reserve(input.num_columns());
+
+    auto& h_dst_buf_info = partition_buf_size_and_dst_buf_info->h_dst_buf_info;
+    auto& h_dst_bufs     = src_and_dst_pointers->h_dst_bufs;
+
+    auto cur_dst_buf_info = h_dst_buf_info;
+    detail::metadata_builder mb(input.num_columns());
+
+    for (std::size_t idx = 0; idx < num_partitions; idx++) {
+      // traverse the buffers and build the columns.
+      cur_dst_buf_info = build_output_columns(input.begin(),
+                                              input.end(),
+                                              cur_dst_buf_info,
+                                              std::back_inserter(cols),
+                                              h_dst_bufs[idx],
+                                              mb);
+
+      // pack the columns
+      result.emplace_back(packed_table{
+        cudf::table_view{cols},
+        packed_columns{std::make_unique<std::vector<uint8_t>>(mb.build()),
+                       std::make_unique<rmm::device_buffer>(std::move(out_buffers[idx]))}});
+
+      cols.clear();
+      mb.clear();
+    }
+
+    return result;
+  }
+
+  cudf::table_view const input;
+  std::size_t const user_buffer_size;
+  rmm::cuda_stream_view const stream;
+  rmm::mr::device_memory_resource* const mr;
+  rmm::mr::device_memory_resource* const temp_mr;
+
+  // whether the table was empty to begin with (0 rows or 0 columns) and should be metadata-only
+  bool const is_empty;
+
+  std::size_t const num_partitions;
+
+  // number of source buffers including children
+  size_type const num_src_bufs;
+
+  // number of source buffers including children * number of splits
+  std::size_t const num_bufs;
+
+  std::unique_ptr<packed_partition_buf_size_and_dst_buf_info> partition_buf_size_and_dst_buf_info;
+
+  std::unique_ptr<packed_src_and_dst_pointers> src_and_dst_pointers;
+
+  //
+  // State around the chunked pattern
+  //
+
+  // chunked_pack will have 1 or more "chunks" to iterate on, defined in chunk_iter_state
+  // contiguous_split will have a single "chunk" in chunk_iter_state, so no iteration.
+  std::unique_ptr<chunk_iteration_state> chunk_iter_state;
+
+  // Two modes are allowed:
+  //  - user provided buffer: as the name implies, the user has provided a buffer that must be at
```

Comment:
In this context a user is not a libcudf user, but rather libcudf code making use of this struct, right? In particular, `contiguous_split` vs `chunked_pack`?

Reply:
That is correct. I will clarify this comment, it was written before we started calling this `chunked_pack`.. before this it was `chunked_contiguous_split` which was all kinds of confusing :)

Reply:
That is correct. I will clarify this comment, it was written before we started calling this `chunked_pack`.. before this it was `chunked_contiguous_split` which was all kinds of confusing :)